### ISSUE TITLE
Patch based Image Augmentor?

### ISSUE THREAD

**Issue by Skylion007** [ppwwyyxx/tensorpack#133]
Is there an augmentor I can use to divide an image into patches and then reassemble them as returned patches? Or would I have to modify the actual network architecture to do so using tf.extract_image_patches and such? I'm just curious what the best way to go about implementing this with Tensorpack would be?

**Comment by ppwwyyxx** [comment]
I don't know what exactly you want, but we certainly don't have anything like this.
If your input->output is one image -> one image, you can write an augmentor, by subclass ImageAugmentor and implement the related methods.
If you want one image -> several images in one tensor, or one datapoint -> one datapoint, you can use MapData to specify a mapping function.
More generally you can write a DataFlow subclass to implement whatever you want to do with your data, as long as the output can be represented as a tensor in the end.

**Comment by Skylion007** [comment]
I just want a dataflow process to make it easy to scale up an algorithm to patch based input. For example, let's say you have a massive 10000x10000 px image (not uncommon for satellite imagery). 

Dataflow would break image into small 256x256 patches. -> (break in batches of batchsize?) -> run network -> reassemble image by concatenating all patches into their proper place. 

^ (Also if patches overlap, it would be great to easily merge them by averaging the overlapping portions.)

The mapdata sounds like it might do what I want it to so I may try that. I was surprised that the repo didn't have any functions to do this since it's incredibly useful for Patch Based GANS.

**Comment by PatWie** [comment]
You can encode this as a part of the graph. (I did it here https://arxiv.org/abs/1607.04433)

The dataflow provided the patches as a stack (tensor of 5 dimensions). It is just transposing and reshaping the data to handle a virtual batchsize of batchsize*patches. However, I am not aware of an easy solution to actually use the combined image again within the loss function. Usually this is trained on patches individually or one uses fully convolutional networks.

But if your 10000x10000px does not fit to the gpu, it won't fit as patches on the gpu.

**Comment by ppwwyyxx** [comment]
@Skylion007  I still don't understand what you mean especially the context. Is this for training as well or only for testing? Doing stuff after getting the output of a network is not a job of DataFlow, but it doesn't sound like anything you would do for training either.

This is not a GAN repo, and I'm not even sure what patch-based GAN means to you. If you are talking about the PatchGAN mentioned in the image to image paper, it is fully convolutional and you can run it on super large images already, without having to split things into patches and building a batch.

**Comment by Skylion007** [comment]
To clarify it doesn't need to use the combined images in the loss function. I just want an easy way to output and reassemble the patches. I am asking because I do not know if the dataflow guarantees ordering within the batches and that they get placed in the same order for the output. 

I was using the GANs as an example. To be exact this is the paper I wanted to implement: http://www.ingentaconnect.com/content/ist/jist/2016/00000060/00000001/art00003 . (See Figure 4).

![image](https://cloud.githubusercontent.com/assets/2053727/22600775/0b90a550-ea0a-11e6-9628-91b4c17d3a53.png)


To be exact here is the figure from the paper that shows the process I want to implement.

I apologize as I just recently got interested in deep learning so I am still learning all the appropriate terminology. I have tried using tf.extract_image_patches(), but it seems to run out of memory when I attempt on larger images in vanilla Tensorflow. I was hoping this repo could automate the process and make it easier to do the preprocessing on the fly. 

I also would like to complement you on your excellent support on this repo. You guys have had absolutely excellent response times and are willing to help people fully utilize the framework. I just wished to applaud you on that front.

**Comment by PatWie** [comment]
Then the definitiv answer is that the data provider needs to [generate patches](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.image.extract_patches_2d.html) and combining by Hanning-Window averaging or similar methods.

**Comment by Skylion007** [comment]
Ah yes, that's what I was thinking. The issue is how to elegantly recombine them. Is there a post-callback I can use? I assume I'll need to collect the patches on the output and hold them in RAM so I was curious if there was a callback or runner that would help me do the patch recombination. (Hold N output values then recombine them perform "some action" on them). Bonus points if I could vary N as needed to compensate for different image sizes (not a requirement by any means, but it would be a nice addition).

TLDR: The issue I have now is how to recombine them elegantly in the Tensorpack framework.

**Comment by ppwwyyxx** [comment]
Thanks for the details and we love to hear different use cases from different people.
From the paper it looks like this is only used for prediction, after a model has been trained? If so it's not about DataFlow or callbacks.
